---
title: "An Introduction To ngsReports"
author:
- name: Christopher Ward
  affiliation: School of Biological Sciences, University of Adelaide
- name: Hien To
  affiliation: Bioinformatics Hub, University of Adelaide
- name: Steve Pederson
  affiliation: Bioinformatics Hub, University of Adelaide
  email: stephen.pederson@adelaide.edu.au
package: ngsReports
output:
  BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{An Introduction To ngsReports}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r}
knitr::opts_chunk$set(message = FALSE, warning = FALSE)
```


# Introduction

The package `ngsReports` is designed to bolt into data processing pipelines and produce combined plots for FastQC reports across an entire set of libraries.
Output as generated by the tool [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/) will be loaded and replotted using functions from `r CRANpkg("ggplot2")` and `r CRANpkg("plotly")`, as well as custom summary plots.
Tables of read counts can also be easily generated.

# Basic Usage

## The default report

In it's simplest form, a default report can be generated simply by specifying a directory containing the output from FastQC and calling the function `writeHtmlReport()`.

```{r}
library(ngsReports)
```


```{r, eval = FALSE}
fileDir <- system.file("extdata", package = "ngsReports")
writeHtmlReport(fileDir)
```

This function will transfer the default template to the provided directory and produce a single `.html` file containing interactive summary plots of any FastQC output found in the directory.
FastQC output can be `*fastqc.zip` files or the same files extracted as individual directories.

The default template is provided as `ngsReports_Fastqc.Rmd` in the package directory.
This can be easily modified and supplied as an alternate template to the above function.

```{r, eval = FALSE}
altTemplate <- file.path("path", "to", "template.Rmd")
writeHtmlReport(fileDir, template = altTemplate)
```

# Advanced Usage

## Classes Defined in the Package

The package `ngsReports` introduces four basic `S4` classes: 

* `FastqcFile` & `FastqcFileList`
* `FastqcData` & `FastqcDataList`

`FastqcFile` and `FastqcData` objects respectively hold the location and information from a single report as generated by the stand-alone tool `FastQC`^[https://www.bioinformatics.babraham.ac.uk/projects/fastqc/].
These are then extended into lists for more than one file.
`FastqcFile` and `FastqcFileList` objects simply contain the path to these reports after checking the files contain the files always output by this tool.
`FastqcData` and `FastqcDataList` objects contain the *parsed data* from the reports.
For most users, the primary class of interest will be the `FastqcDataList`

## Loading FastQC Data Into `R`

To load a set of `FastQC` reports into `R` as a `FastqcDataList`, specify the set of files, then call the function `getFastqcData()`.

```{r}
fileDir <- system.file("extdata", package = "ngsReports")
files <- list.files(fileDir, pattern = "fastqc.zip$", full.names = TRUE)
fdl <- getFastqcData(files)
```

From here, all FastQC modules can be obtained as a `tibble` (i.e. `data.frame`) using the functions:

* `Basic_Statistics(fdl)`
* `Per_base_sequence_quality(fdl)`
* `Per_sequence_quality_scores(fdl)`
* `Per_base_sequence_content(fdl)`
* `Per_sequence_GC_content(fdl)`
* `Per_base_N_content(fdl)`
* `Sequence_Length_Distribution(fdl)`
* `Sequence_Duplication_Levels(fdl)`
* `Overrepresented_sequences(fdl)`
* `Adapter_Content(fdl)`
* `Kmer_Content(fdl)`
* `Per_tile_sequence_quality(fdl)`

In addition, the read totals for each file in the library can be obtained using, whiich can be easily used to make a table of read totals.

```{r, results='hide'}
reads <- readTotals(fdl)
```

```{r}
library(dplyr)
library(pander)
filter(reads, grepl("R1", Filename)) %>% 
  pander(big.mark = ",")
```


One additional module is available and taken directly from the text within the supplied reports

* `Total_Duplicated_Percentage(fdl)`

## Generating Plots

Plots created from a single `FastqcData` object will resemble those generated by the `FastQC` tool, whilst those created from a `FastqcDataList` will be combined summaries across a library of files.
All plots are able to be generated as interactive plots using the argument `usePlotly = TRUE`.

Most modules have been enabled for plotting using default `S4` dispatch.

### plotSummary()

A summary of all `PASS/WARN/FAIL` flags can be simply generated

```{r plotSummary, fig.cap="Default summary of FastQC flags.", fig.wide = TRUE}
plotSummary(fdl)
```

### plotReadTotals()

Read totals from all files in the library can be simply plotted.
By default, the number of duplicated sequences from the `Total_Duplicated_Percentage` module are shown, but this can be disabled by setting `duplicated = FALSE`.

```{r}
plotReadTotals(fdl)
```

As these are `ggplot2` objects, the output can be modified easily using conventional `ggplot2` syntax.

```{r}
plotReadTotals(fdl) +
  geom_vline(xintercept = 25000, linetype = 2) 
```

### plotBaseQualities()

A heatmap of the entire `Per base sequence quality` scores can by generated, using the mean as representative of the scores at each position.
The median can also be selected using `plotValue = "Median"`

```{r}
plotBaseQualities(fdl)
```

The conventional boxplot for a single file can be created by specifying a single `FastqcData` or `FastqcFile` object.

```{r}
plotBaseQualities(fdl[[1]])
```

Boxplots of any combinations can also be drawn from a `FastqcDataList` or `FastqcFileList` by setting the argument `plotType = "boxplot"`

```{r}
plotBaseQualities(fdl[1:4], plotType = "boxplot")
```

### plotSequenceQualities()

A heatmap of mean sequence qualities can be generated, with the alternative being a set of overlaid lines.

```{r}
plotSequenceQualities(fdl)
```

```{r}
r2 <- grepl("R2", fileName(fdl))
plotSequenceQualities(fdl[r2], plotType = "line")
```

### plotSequenceContent()

Heatmaps showing a composite colour for each position can be drawn. 
Colours are combined using `rgb()` with `A`,`C`, `G` and `T` being represented by green, blue, black and red respectively.

```{r}
plotSequenceContent(fdl)
```

Individual plots can also be drawn using a `FastqcData` or `FastqcFile` object

```{r}
plotSequenceContent(fdl[[1]])
```

### plotNContent()

N content can also be checked.
The example dataset contained no N content, and so a "blank" plot will be produced.

```{r}
plotNContent(fdl)
```

### plotSequenceLengthDistribution

```{r}
plotSequenceLengthDistribution(fdl)
```

### plotDuplicationLevels()

```{r}
plotDuplicationLevels(fdl)
```

### plotAdapterContent()

Total Adapter Content is able to be plotted for complete set of files

```{r}
plotAdapterContent(fdl)
```

Adapter content by individual adapter type is available for each individual file, as identified by FastQC.

```{r}
plotAdapterContent(fdl[[1]])
```

### plotKmers()

Althought not highly informative, particularly considering the binning across positions by FastQC, the totals of identified Kmers are plotted for a set of files.

```{r}
plotKmers(fdl)
```

The top 6 Kmers, as conventionally displayed by FastQC can be simply obtained, however Kmers which overlap at indiviual positions may not be visible.

```{r}
plotKmers(fdl[[1]], axis.text.x = element_text(angle = 90)) 
```

## GC content

### The class TheoreticalGC

A selection of Theoretical GC content is supplied with the package in the object `gcTheoretical`, which has been defined with the additional `S4` class `TheoreticalGC`.
GC content was calculated using scripts obtained from https://github.com/mikelove/fastqcTheoreticalGC.
Available genomes and transcriptomes can be obtained using the functions `genomes()` and `transcriptomes()`

```{r, eval=FALSE}
transcriptomes(gcTheoretical)
genomes(gcTheoretical)
```

### plotGcContent()

This function offers numerous possibilities, and is able to be presented as the difference from a theoretical GC content.
By default, the difference is shown from the `Hsapiens` *genome*


```{r}
plotGcContent(fdl)
```

This can be changed to the transcriptome by setting `theoreticalType = "Transcriptome"`.
Alternative species can be also be selected from those made available

```{r}
plotGcContent(fdl, theoreticalType = "Transcriptome", species = "Mmusculus")
```

Customized theoretical GC content can be generated using input DNA sequences from a fasta file

```{r,message=FALSE,warning=FALSE}
faFile <- system.file("extdata", "Athaliana.TAIR10.tRNA.fasta", package="ngsReports")
plotGcContent(fdl, Fastafile = faFile)
```

Line plots can also be produced.

```{r}
plotGcContent(fdl, plotType = "line",  theoreticalType = "Transcriptome")
```

## Overrepresented Sequences

Instead of identifying common sequences across a set of libraries, overrepresented sequenes are summarised by their possible source as defined by FastQC.

```{r}
plotOverrepresentedSummary(fdl)
```

The top `n`can be plotted for an individual file, again broken down by their possible source, and coloured based on their `WARN/FAIL` status

```{r, fig.wide = TRUE}
plotOverrepresentedSummary(fdl[[1]])
```

In addition to the above, overrepresented sequences can be exported as a FASTA file for easy submission to `blast`.

```{r, eval = FALSE}
exportOverrepresented(fdl, n = 10)
```


# Session info {.unnumbered}

```{r sessionInfo, echo=FALSE}
sessionInfo()
```
